sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,304 @@ def transform(self, X):
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing > 0):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing*100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = _get_mask(X, self.missing_values)
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            neighbors = self._fitted_neighbors.kneighbors(
+                X[row_has_missing, :], n_neighbors=n_neighbors_new)
+
+            # Get row index, distance, and weights of donors
+            knn_distances, knn_row_index = neighbors
+            # Remove self from list of donors
+            if n_neighbors_new > self.n_neighbors:
+                row_index = np.arange(X.shape[0]).reshape((X.shape[0], 1))
+                row_index = row_index[row_has_missing, :]
+                not_duplicate_index = np.where(~(row_index == knn_row_index))
+                knn_row_index = knn_row_index[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+                knn_distances = knn_distances[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+            weights = _get_weights(knn_distances, self.weights)
+
+            # Vertically split sets of k-donor indices and repeat each set by
+            # missing count in the corresponding recipient row
+            knn_row_index = np.vsplit(knn_row_index, knn_row_index.shape[0])
+            row_repeats = row_total_missing[row_total_missing != 0]
+            knn_row_index = np.repeat(
+                knn_row_index, row_repeats, axis=0).ravel()
+
+            # Get column index of donors
+            row_missing_index, col_missing_index = np.where(mask)
+            knn_col_index = np.repeat(col_missing_index, self.n_neighbors)
+
+            # Calculate kNN score and impute
+            donors = fitted_X[
```

Comment:
This is quite a neat way to do it :)

Reply:
wait what? is this really happening? ... lol thanks!

Reply:
I know, I need to work on my positive feedback.

Reply:
haha was just messing with you. plus i don't blame you, if i was reviewing endless lines of code and then suddenly i have to deal with ppl coming up with ad-hoc concocted definitions of "future-proofing", i don't think i would be too liberal with positive feedbacks either :)

Reply:
Yes, but I have also been a contributor to other projects, and know the difference between being received with business-as-usual and being received with joy. I just struggle to reflect that in my own reviewing.

Reply:
i see where you are coming from. but i can't imagine it not being challenging for anybody if they were handling the amount of work you usually do here, and not to mention having to interact and coordinate with people from the entire spectrum of technical abilities, cultures, languages, etc.